Implementation of `ParquetFile.append_as_row_groups()`

yohplala opened the pull request:
'Re-ordering' of 'write' to separate 'append' part and make it available in ParquetFile through 'append_as_row_groups' method.

As per initial PR:

__Implementation of `ParquetFile.append_as_row_groups()`__

This PR completes the proposal made in ticket #676.

The work done consisted in extracting from current `writer.write()` code dealing
  - with 'multi-part' format into a `write_multi()` (similar to the existing `write_simple()`)
  - with data index management (to have it fit for parquet format) into a `reset_row_idx()`
  - with some checks related 'append' into `ParquetFile.append_as_row_groups()`

In the end, the root work (appending or writing a new file) is achieved either by `write_multi` or `write_simple`.

Both `writer.write` and `api.ParquetFile.append_as_row_groups` call `write_multi` or `write_simple`:
- `writer.write` call them directly when writing a file from scratch
- `api.ParquetFile.append_as_row_groups` call them when `append=True` (or overwrite)

For appending, `writer.write.py` calls `api.ParquetFile.append_as_row_groups`, bypassing creation of metadata (which was not the case before, but was a step uselessly done (these metadata not being used afterwards).

__Among others changes:__
- I indicated in `writer.write` docstring the mention: 'Ignored if appending to an existing parquet dataset.' when related parameter is not used when appending.
- several test cases have been added in `test_api.py` to check `api.ParquetFile.append_as_row_groups`, and complete case when `append='overwrite'`.

yohplala commented:
Hi @martindurant 
Please, do you have any feedbacks regarding this PR?

martindurant commented on fastparquet/api.py:
```diff
@@ -342,7 +343,7 @@ def iter_row_groups(self, filters=None, **kwargs):
                 yield df
 
-    def remove_row_groups(self, rgs, write_fmd:bool = True,
+    def remove_row_groups(self, rgs, write_fmd:bool=True,
```
Please revert format-only changes

yohplala replied:
done

martindurant commented on fastparquet/api.py:
```diff
@@ -372,11 +373,11 @@ def remove_row_groups(self, rgs, write_fmd:bool = True,
                 remove_with = default_remove
         if not isinstance(rgs, list):
             rgs = [rgs]
-        rgs_to_remove = row_groups_map(rgs)
```
please revert this rename to make my life easier

yohplala replied:
done

martindurant commented on fastparquet/api.py:
```diff
@@ -391,10 +392,94 @@ def remove_row_groups(self, rgs, write_fmd:bool = True,
+                             open_with=default_open, mkdirs=None, append=True,
+                             stats=True):
+        """
+        Append data as new row groups to disk. Updated `ParquetFile` metadata
```
Top of docstring should be kept to one line

martindurant replied:
Reminds me: how is this method different from `write(..., append=True)` ?

yohplala replied:
The differences with `write` are:
- it is used on an already instanciated `ParquetFile`, (if the instance is already existing/loaded, you don't re-load it again as opposed to `write`)
- there are fewer checks on the `ParquetFile` instance / it requires less parameters, as some of them are defined directly from the instance attributes (`partition_on` for instance)
- it leaves the option to write or not common metadata to disk (`write_fmd=False`)

After this PR, I would like to prepare a new one containing documentation and an illustrative test case. 
This will hopefully illustrate better the interest of completing the set of utilities to manage row groups of a `ParquetFile` instance.

This documentation would be a kind of 'tutorial' / example about how using these functions together to 
update a parquet dataset.

Assuming:
- the user defines a way to identify _overlaps_ in row groups.
- the user defines a way to identify _duplicates_ in rows.

Then the following _update methodology_ with fastparquet is possible:
```python
# Retrieve overlaps between existing row group and new data.
rgs_overlaps = filter_row_groups()

# Concat overlapping row groups with new data, and drop duplicates in rows.
df_to_add = pd.concat(new_data, pf[rgs_overlaps].to_pandas()).drop_duplicates()

# Remove overlaps from disk.
pf.remove_row_groups(rgs_overlaps, write_fmd=False)

# Add new data, along with data from overlapping row groups.
pf.append_as_row_groups(df_to_add)

# Sort row groups if relevant.
pf.fmd.row_groups = sorted(pf.row_groups, key=...)

# Finally write common metadata.
pf._write_common_metadata()
```
I would like to provide 2 application test cases:
- one with overlaps defined by partition. For instance:
`filter_row_groups(pd, filter=[('name', 'in', ['Yoh','Fred'])`
- the other one with overlaps defined by an ordered column. for instance a time series:
`filter_row_groups(pd, filter=[('timestamp', '>=', pd.Timestamp('2021/01/01'), ('timestamp', '<=', pd.Timestamp('2021/01/04')])`
and then derive the proposed procedure.

Compared to the `overwrite`, this approach:
- pros:
  - does not force the user to rely on partitions.
  - does not the limit the user to have a single row group per partitions
- cons:
  - requires more knowledge about these different row group utilities (hence I think a documentation page is worthwhile)
  - (similar) more lines of codes


I introduced this in #676, I am sorry if I am writing a lot of things not necessarily very clear.

yohplala replied:
(made some updates to complete above answer)

yohplala replied:
Hi @martindurant 
Replying here to
_By the way, is "insert" possible (i.e., shuffling the other row-groups and filenames as necessary)?_

I rephrased this question in my mind into:
_Could your work be extended so as to make insertion of row groups ('shuffling' other row-groups and filenames as necessary?_

To this question, my answer is yes, and I think it is actually a good idea!
It has not been my target so far (as it is possible to do it by using `sorted` on row group list directly), but I think doing insertion directly in `append_as_row_groups` (that I will rename into `write_row_groups` as per our above discussion) is a neater solution.

I think i would also change the interface of `write_row_groups()` so that it accepts an iterable of dataframes (e.g. list of dataframes or iterator of dataframes, each defining a row group) instead of a single dataframe to be split. for each of these row groups, the list of indexes where these row groups have to be inserted will have to be provided optionally (if not provided, it will be usual append).

With these changes, we can make the existing `overwrite` feature a separate function, similar in the way it is 'external' as the existing `merge`, and as efficient as it is now (but which implementation is intricated into `write`).

As I see it, it will thus bring us more modular code, easier to read and maintain, so yes, I am keen on investing this!
I will work out something in the coming days, and push updates, probably by the end of this week.
Thanks for your constructive insights!
Bests,

martindurant commented on fastparquet/api.py:
```diff
@@ -391,10 +392,94 @@ def remove_row_groups(self, rgs, write_fmd:bool = True,
         if write_fmd:
             self._write_common_metadata(open_with)
 
+    def append_as_row_groups(self, data, row_group_offsets=None,
```
Appends *or* overwrites

yohplala replied:
If it is acceptable, I would propose to keep the 'append' only, having in mind we can set the `append` parameter to either `True`, `False` or `overwrite`.

martindurant replied:
I mean, the process invoked by this method does not match its name. People would be surprised to call `append_as_row_groups` and loose some data whatever other arguments they may be passing.

yohplala replied:
Ok,
I can propose `write_as_row_groups()` maybe?
(I am really not sure about a `append_as_or_overwrite_row_groups`, but this is certainly the farther we can go into introducing what it can do)

Maybe `set_as_row_groups` is also fitting.

martindurant replied:
`write_row_group` ?

By the way, is "insert" possible (i.e., shuffling the other row-groups and filenames as necessary)?

yohplala replied:
Ok for me for `write_row_group` but maybe with a `s`? (`write_row_groups`)

I don't understand the 2nd part of the comment.
Without fully understanding it, I would say 'it depends'. It depends on the mode retained in 'append' just as in current case:
- if append is True, row groups are appended
- if append is 'overwrite', ... well, it is not exactly insertion,  but replacement. This is a specific case with partitions.

Insertion of row groups is 'mimicked' in the procedure I have discussed below.
So, I would say 'yes', row groups can be shuffled thanks to `sorted`.
If I am not answering correctly, and if the discussion is related to to the 'update' procedure detailed below, I would propose to go on the discussion in the corresponding section below if this makes sense.
I should be able to answer other points raised on Wednesday evening.

yohplala replied:
Renaming to `write_row_groups()` done.

martindurant commented on fastparquet/api.py:
```diff
@@ -391,10 +392,94 @@ def remove_row_groups(self, rgs, write_fmd:bool = True,
+                        open_with=open_with, mkdirs=mkdirs,
+                        partition_on=partition_on, append=append, stats=stats)
+        self._set_attrs()
+        return
```
return unnecessary

yohplala replied:
removed

martindurant commented on fastparquet/test/test_overwrite.py:
```diff
@@ -10,16 +10,14 @@
                         'location': ['Paris', 'Paris', 'Milan'],
                         'color': ['red', 'black', 'blue']})
-    write(tempdir, df1, row_group_offsets=0, file_scheme='hive',
+    write(tempdir, df1, row_group_offsets=[0], file_scheme='hive',
```
Why did this change?

yohplala replied:
undone

martindurant commented on fastparquet/util.py:
```diff
@@ -120,6 +120,58 @@ def check_column_names(columns, *args):
 
+def reset_row_idx(data: pd.DataFrame) -> pd.DataFrame:
+    """
+    Shift row index to columns of the DataFrame, compatible for storing to
```
I don't understand this description

yohplala replied:
I updated it into:
```python
    Reset row index as a column of the DataFrame.
    Multi-index are stored in columns, one per index level.
```

martindurant replied:
You need to have a single line description and then a blank line

martindurant commented on fastparquet/writer.py:
```diff
@@ -762,22 +764,44 @@ def make_metadata(data, has_nulls=True, ignore_columns=None, fixed_text=None,
 
 
 def write_simple(fn, data, fmd, row_group_offsets, compression,
-                 open_with, has_nulls, append=False, stats=True):
```
Changing this signature, even if the argument is not used, might mess up other libraries that call it.

yohplala replied:
Ok, I have added it back.
It was raising no error in pandas or dask test cases.
For my information, what should be the steps to operate such a change? (I guess at some point, cleaning a function signature makes sense to maintain a library?)

martindurant replied:
Ideally, we would set it to `=None` and raise PendingDeprecationWarning if it has a non-None value.

yohplala replied:
`PendingDeprecationWarning` is set for
- `has_nulls` in `writer.write_simple()`,
- `sep` in `api.ParquetFile.__init__()`

martindurant replied:
Maybe remove these just for the moment, make this a separate PR (and perhaps after the next release)

martindurant commented on fastparquet/writer.py:
```diff
@@ -891,154 +1052,101 @@ def write(filename, data, row_group_offsets=50000000,
         For object columns, this gives the data type, so that the values can
-        be encoded to bytes. Possible values are bytes|utf8|json|bson|bool|int|int32|decimal,
+        be encoded to bytes.
+        Possible values are bytes|utf8|json|bson|bool|int|int32|decimal,
```
Please revert formatting

yohplala replied:
done

martindurant reviewed (commented):
Thoughts so far, without having yet looked through the big refactor on `write()`

yohplala commented:
Hi, here is a status of the on-going work.

###Content of last commit
The 'row group insertion + sorting' is (obviously) closely connected to the 'append=overwrite' mode, and it seemed necessary to me to refactor it while equipping `pf.write_row_groups()` with the 'row group sorting' feature.

It seems to me now that the name of this PR is kind of a tree that hides the forest. It is actually a refactoring of `write` function, which has been split between `write_multi`, `pf.write_row_groups`, `iter_dataframe`, and `update`.

Features of `pf.write_row_groups` over `write` are now:
  - uses as input a ParquetFile instance (if the instance is already existing/loaded, you don't re-load it again as opposed to write),
  - returns the modified ParquetFile instance (if you need it for subsequent use, no need to reload it)
  - fewer checks on the ParquetFile instance / requires less parameters, as some of them are defined directly from the instance attributes (partition_on for instance)
  - leaves the option to write or not common metadata to disk (write_fmd=False)
  - allow use of a key to sort row groups
  - requires an iterable of dataframes (aka row groups as separate dataframes)
 
###Filename shuffling
This PR is not yet completed, as filename shuffling is still missing.
For this feature, my proposal is to optionally make possible renaming of filenames as per related row group position in row group list. My target is to finalize this by end of this week.
It should not be too much code.

###`update()`
A comment regarding new `update` function.
This new function manages the 'overwrite' mode.
Its name is currently inadequate, but my next step (next PR) will be to make it the skeleton to be used for operating the update of an existing dataset.
Its content relates to what I was formalizing when speaking about _overlaps_ between row-groups and _duplicates_ between rows.

So as not to change its name in the next PR, I already name it `update`.
If this is too much inadequate, its name as per its current scope, would be `overwrite_partitioned`.
Please, let me know if i should change it or not.

###Iterable of dataframes for iterating separate row groups
Some other frameworks featuring out of core dataframe management (for instance vaex) can generate pandas dataframe in chunks (to manage data that do not fit into RAM).
`pf.write_row_groups()`, `write_multi` and `write_simple` now accepts directly such an iterable.
To benefit from fastparquet efficiency to write parquet file from vaex dataframe, user can now take advantage of this feature.

yohplala commented:
@martindurant 
I am working on the filename renaming.
I am wondering how is set `rg.columns[0].file_path`?

Is it set at parquet files discovery when instanciating a `ParquetFile`?
If yes, then changing the name of the file is enough.

But if this attribute also persists in the metadata, I need to update it there as well.
In this latter case,I don't know how. Please could you provide some guidance?

Thanks for your help, bests

yohplala commented:
> But if this attribute also persists in the metadata, I need to update it there as well. In this latter case,I don't know how. Please could you provide some guidance?

Ok, found it:
- yes, it is persisted
- code example in `write_muli`:
```python
            partname = join_path(dn, part)
            with open_with(partname, 'wb') as f2:
                rg = make_part_file(f2, row_group, fmd.schema,
                                    compression=compression, fmd=fmd,
                                    stats=stats)
            for chunk in rg.columns:
                chunk.file_path = part
```

martindurant commented:
As usual, it isn't simple :)
A _metadata file contains a full thrift description of the dataset, including the file_paths of all of the columns of all of the row-groups. If this special file exists, fastparquet will use it and not search for files or even read their own copies of the matadata. Note that the file_paths in the metadata of the file that contains those columns should be None, it's only used when referencing a column chunk in *another* file.

If the special file does not exist, fastparquet effectively constructs the corresponding thrift object in memory.

Other frameworks do something different, including fastparquet-on-dask. They may well ignore any _metadata file even if it exists, so here the filenames and their ordering becomes important.

yohplala commented:
> As usual, it isn't simple :) A _metadata file contains a full thrift description of the dataset, including the file_paths of all of the columns of all of the row-groups. If this special file exists, fastparquet will use it and not search for files or even read their own copies of the matadata. Note that the file_paths in the metadata of the file that contains those columns should be None, it's only used when referencing a column chunk in _another_ file.

Ok, if think it fits with what I am implementing.
I will update `file_path` in each `row_group` in `fmd.row_groups`.

martindurant commented:
Sorry, I created some merge conflicts for you

yohplala commented:
> Sorry, I created some merge conflicts for you

Martin, it will be that is sorry when I will finally pull-request this 800+ lines of code.

yohplala commented:
Hi @martindurant 

Overall, I am done with this PR, except for this `file_path` that dask wants as `string`.
I could put some `if` here and there, but even in dask itself, there is a function that I canno tmodify:

```python
dask/dask/utils.py:1335: in natural_sort_key
    return [int(part) if part.isdigit() else part for part in re.split(r"(\d+)", s)]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

pattern = '(\\d+)', string = b'a=apple/b=large/part.0.parquet', maxsplit = 0
flags = 0
    def split(pattern, string, maxsplit=0, flags=0):
        """Split the source string by the occurrences of the pattern,
        returning a list containing the resulting substrings.  If
        capturing parentheses are used in pattern, then the text of all
        groups in the pattern are also returned as part of the resulting
        list.  If maxsplit is nonzero, at most maxsplit splits occur,
        and the remainder of the string is returned as the final element
        of the list."""
>       return _compile(pattern, flags).split(string, maxsplit)
E       TypeError: cannot use a string pattern on a bytes-like object
```
Hmmm, I am running out of ideas.
If you have any on what should be done, I would gladly read about them.

I had in mind to have bytes only in metadata, but maybe we should do the reverse.
Once `file_path` is written, we revert it to `string`. I have to check where this is done.
Then, we only keep string everywhere, except in this function that actually writes the value.
Do you think this could be ok? (stopping here for today)

yohplala commented:
Hi @martindurant 
It now seems ok.
So as I could conclude yesterday, I let `pf.fmd;row_groups[].columns[].file_path` be a string as soon as it is set.

I think that with everything that I could write, everything has been 'introduced' so to say.

One thing remaining regarding `update` that is the refactoring of `write(...., append='overwrite')`.
With the formalization I have made, it removes the constraint to use a single row group per partition.
The only constraints that remain when using `update` (or the redirection `write(...., append='overwrite')`) is that the existing dataset has to be written with partitions (which also implicitly implies `hive` format).

Please, let me know about your feedback.
Again I am sorry for such a big PR.
Next ones will be much more focused, I will take care about this.

yohplala commented:
Hi @martindurant ,
Please, do you have any feedbacks for this PR?

martindurant commented on fastparquet/writer.py:
```diff
@@ -930,197 +1078,155 @@ def write(filename, data, row_group_offsets=50000000,
-    nums = [int(match.groupdict()['i']) for match in matches if match]
-    if nums:
-        return max(nums) + 1
+    pids = part_ids(row_groups)
```
Ah, the source of the extra utility function. Is it really useful to pull it out of here, does it get called anywhere else?

yohplala replied:
It is called in `ParquetFile._sort_part_names()`, where it uses the complementary info that are kept in the returned `dict`.

But I do agree, it seems to me we have several functions to handle partition info:
- directory path (partition values)
- file path (partition file id)
To query several type of info:
- list of partition file ids
- list of partition values
- max partition file id
- number of row groups per partition files
- mapping between partition file ids and row group id of the 1st row group the file contains

At present, it seems to me most of these are in `api.py`.

martindurant commented on fastparquet/writer.py:
```diff
@@ -1243,11 +1351,120 @@ def merge(file_list, verify_schema=True, open_with=default_open,
+def update(dirpath, data, row_group_offsets=None, sort_pnames:bool=True,
+           compression=None, open_with=default_open, mkdirs=None, rename=None,
+           remove_with=None, stats=True):
+    """Merge new data to existing parquet dataset.
```
So what exactly is the correspondence between this and the new methods on ParquetFile? Do we need both?

yohplala replied:
`ParquetFile.write_row_groups()` 'only':
- adds new data to an existing dataset,
- define where to insert new row groups in row group list,
- can reorder partition files (rename them),
- modify metadata

`update` is a 'macro' function that rolls out currently only one logic for updating a partitioned dataset (I identify a different update logic in the case of an ordered dataset, having no partition). It is the re-implementation of `append=overwrite` feature.
- it identifies partition files to be removed based on partition values identified in new data,
- it calls `ParquetFile.remove_row_groups()` to remove them
- it defines the logic to sort row groups that will be added from new data
- it calls `ParquetFile.write_row_groups()` to add new data (and `ParquetFile.write_row_groups()` 

I would not say there is correspondence between both.
- `update` is the implementation of an update logic for partitioned datasets.
- `ParquetFile.write_row_groups` 'only' add new row groups to an existing dataset.

martindurant reviewed (commented):
Reviewed as far as writer.py, which only has a couple of comments. As usual, the extensive refactor there makes it tricky to see what has really changed and what not.

yohplala commented on fastparquet/api.py:
```diff
@@ -186,14 +187,14 @@ def _parse_header(self, f, verify=True):
-        for rg in fmd[4]:
-            chunks = rg[1]
+        for rg in fmd.row_groups:
+            chunks = rg.columns
```
Ok, done. I let the readable version in comment. It seems to me it is good to keep track of what are doing these lines of code but I can remove. Please, let me know.

yohplala commented on fastparquet/api.py:
```diff
@@ -239,7 +240,10 @@ def basepath(self):
 
     def _read_partitions(self):
-        paths = [rg[1][0].get(1, "") for rg in self.row_groups if rg[1]]
+        paths = [rg.columns[0].file_path
```
Done. Same I let some comments, please let me know if I should remove.

yohplala commented on fastparquet/api.py:
```diff
@@ -278,11 +282,12 @@ def __getitem__(self, item):
 
     def row_group_filename(self, rg):
         if rg.columns and rg.columns[0].file_path:
+            fpath = rg.columns[0].file_path
```
Sorry, I have removed.

yohplala commented on fastparquet/api.py:
```diff
@@ -371,6 +376,15 @@ def remove_row_groups(self, rgs, write_fmd:bool = True,
             a .fs file system attribute
         """
+        if not isinstance(rgs, list):
+            if not hasattr(rgs, '__iter__'):
```
Thanks, I rewrote the code.
I did not keep the `try/except` though, because lines of code immediately after take care of empty row-groups list.
This `isinstance/hasattr` has been here to catch row group list given as generator.

I left a comment explaining why I swith from `[]` (previous code) to `list()`.
- On a generator, `[]` encapsulate it in a list. When you are iterating over the list content, you get the generator, not the items of the generator.
- `list()` materializes the generator into a list. when you iterate over it, it will iterate the row groups.

martindurant replied:
Does the first branch ever get called? A tuple also has `__iter__`, actually many things do. If what you really want to check for generator, why not do exactly that (`inspect.isgenerator`) ?

yohplala commented on fastparquet/api.py:
```diff
@@ -403,10 +415,150 @@ def remove_row_groups(self, rgs, write_fmd:bool = True,
+            If int, row-groups will be approximately this many rows, rounded down
+            to make row groups about the same size;
+            If a list, the explicit index values to start new row groups;
+            If `None`, set to 50_000_000.
```
It is, but it is not set in `writer.write()`.
I moved the code slicing the pandas dataframe in chunks into a separate function: `writer.iter_dataframe()`.
It is where default value is set (the 2 first lines of the function). I could make it a 'constant' and set it at the beginning of `writer.py` though?
```python
import ...

ROW_GROUP_SIZE = 50_000_000

...

def iter_dataframe():
    if row_group_offsets is None:
        row_group_offsets = ROW_GROUP_SIZE
```
What do you think?

martindurant replied:
Actually yes, I like using a constant. Then a user could change it once and have it be the default for all future writes.

yohplala commented on fastparquet/api.py:
```diff
@@ -403,10 +415,150 @@ def remove_row_groups(self, rgs, write_fmd:bool = True,
+            When called with a path/URL, creates any necessary dictionaries to
+            make that location writable, e.g., ``os.makedirs``. This is not
+            necessary if using the simple file scheme.
+        rename : function
```
I agree with this assumption, but fail to see what I should do differently?
I have copied the change you proposed for `remove_with` when we worked on `ParquetFile.remove_row_groups()`:
```python
        if rename is None:
            if hasattr(self, 'fs'):
                rename = self.fs.rename
            else:
                rename = default_rename
```
Should I replace `rename()` by `self.fs.rename()` everywhere I am calling `rename()`?

martindurant replied:
I imagine if you used `self.fs.rename` without the check, you will see no errors, and then you don't need the `rename` argument at all. Actually looking now, I don't thing `.fs` is set for the `ParquetFile(list-of-files)` case, but should be.

It maybe would be good to have `.fs` be an attribute which raises a suitable deprecation warning if we don't have an fs. In another PR.

One outstanding situation where it wouldn't work, is where `ParquetFile` was opened with an already-open file; but in that case, you can't make any modifications anyway.

yohplala commented on fastparquet/api.py:
```diff
@@ -403,10 +415,150 @@ def remove_row_groups(self, rgs, write_fmd:bool = True,
+        if isinstance(data, pd.DataFrame):
+            self_cols = sorted(self.columns + partition_on)
+            if self_cols != sorted(data.columns):
+                # TODO
```
Ok done.
Yet the error message is tested in 2 test cases, which then fail. I updated them as well. (sometime, I do 'try' to minimize the number of changes in my PRs, this was such an attempt ;))
 - `test_output.test_append_fail_incompatible()`
 - `test_api.test_write_rgs_simple_schema_exception()`

yohplala commented on fastparquet/api.py:
```diff
@@ -403,10 +415,150 @@ def remove_row_groups(self, rgs, write_fmd:bool = True,
+                self._write_common_metadata(open_with)
+        self._set_attrs()
+
+    def _sort_part_names(self, write_fmd:bool=True, open_with=default_open,
```
`_sort_part_names()` does modify metadata by modifying `file_path` when files are renamed.

Initially, I did call it with `write_fmd=True` from `ParquetFile.write_row_groups()`.
But then I realized that this function (`_sort_part_names()`) is not necessarily run every time by `ParquetFile.write_row_groups()` (this is user choice); So I could not rely on it to write `fmd` to disk.

I agree with your statement, hence the `_` in the name: `_sort_part_names()`.
I can move it out of `ParquetFile` class (removing the call to `_write_common_metadata`, but the signature will need to include more parameters from `ParquetFile`:
- `self.fs`
- `self.fmd.row_groups`
- `self.basepath`

I made the choice to keep it inside `ParquetFile` class in view of this.
Please, confirm me if I should move it out, I will do.

martindurant replied:
OK, I see what you mean. This is getting to be an extremely long class. When we do eventually tidy up the code, it might be worth splitting into `ParquetFile(ParquetFileReader, ParquetFileModifier)` or something like that.

yohplala commented on fastparquet/api.py:
```diff
@@ -1263,3 +1415,45 @@ def row_groups_map(rgs: list) -> dict:
+    In case of files with multiple row groups, the position (index in row group
+    list) of the 1st group only is kept.
+    """
+    s = re.compile(r'.*part.(?P<i>[\d]+).parquet$')
```
Ok, done.
It is now defined at the beginning of `api.py`:
```python
# Find in names of partition files the integer matching "**part.*.parquet",
# as 'i'.
PART_ID = re.compile(r'.*part.(?P<i>[\d]+).parquet$')
```

yohplala commented on fastparquet/api.py:
```diff
@@ -1263,3 +1415,45 @@ def row_groups_map(rgs: list) -> dict:
+                else f_path.rsplit('/',1)[0])
+
+
+def part_ids(row_groups) -> dict:
```
It is based on `find_max_part()`, but return the list of partition files as a dict with:
- keys as partition file ID
- values as tuples `row_group_id, part_name`.
All this information are re-used in `ParquetFile._sort_part_names()`.
It is also used for `find_max_part` which apply a `max()` on the keys.
(I wanted to remove `find_max_part` but it is directly called in a dask test case)

> Also, I note it will fail if any filenames do not match the given pattern.

You are right, I corrected the code as it was in `find_max_part` to have it not failing if  pattern is not matched.

yohplala commented on fastparquet/cencoding.pyx:
```diff
@@ -863,7 +863,8 @@ def dict_eq(d1, d2):
                 return False
-            if any(a != b for a, b in zip(d1[k], d2[k])):
+            if any(not dict_eq(a,b) if isinstance(a, dict) else (a != b)
+                   for a, b in zip(d1[k], d2[k])):
```
Ok, I have added the link to the discussion as a comment.

yohplala commented on fastparquet/core.py:
```diff
@@ -612,12 +612,12 @@ def read_row_group(file, rg, columns, categories, schema_helper, cats,
         if cat not in assign:
             # do no need to have partition columns in output
             continue
+        fpath = rg.columns[0].file_path
```
Sorry, I reverted the change.

martindurant replied:
(the comment was more for my own records rather than a request for a change, but either way is fine)

yohplala commented on fastparquet/test/test_api.py:
```diff
@@ -1257,3 +1257,127 @@ def test_remove_rgs_simple_merge(tempdir):
     assert pf.to_pandas().equals(df_ref) 
+
+
+def test_write_rgs_simple(tempdir):
```
ahah, at least something good! :)

yohplala commented on fastparquet/test/test_output.py:
```diff
@@ -635,7 +635,8 @@ def test_auto_null_object(tempdir, pnull):
     tmp = str(tempdir)
     cats = np.arange(n)
-    codes = np.random.randint(0, n, size=1000000)
+    rng = np.random.default_rng(4)
```
I confirm this test case has been failing after unrelated changes in the code. I made the update with a seed as we discussed it previously indeed.
I added a link to our discussion back then.

yohplala commented on fastparquet/util.py:
```diff
@@ -193,9 +220,7 @@ def metadata_from_many(file_list, verify_schema=False, open_with=default_open,
-                        chunk.file_path = '/'.join(
-                            [fn, chunk.file_path if isinstance(chunk.file_path, str) else chunk.file_path.decode()]
-                        )
+                        chunk.file_path = '/'.join([fn, chunk.file_path])
```
I changed at `ParquetFile` instantiation the decoding to string behavior.
In current version of fastparquet, only the `file_path` of the 1st column is decoded to string.
This is seen in `_parse_header()` (current version below).
```python
        for rg in fmd[4]:
            chunks = rg[1]
            if chunks:
                chunk = chunks[0]
                s = chunk.get(1)
                if s:
                    chunk[1] = s.decode()
```
Now, included in this PR, all columns are dealt with:
```python
        # for rg in fmd.row_groups:
        for rg in fmd[4]:
            # chunks = rg.columns
            chunks = rg[1]    
            if chunks:
                for chunk in chunks:
                    # s = chunk.file_path
                    s = chunk.get(1)
                    if s:
                        # chunk.file_path = s.decode()
                        chunk[1] = s.decode()
```
To be noticed, there was a related `TODO` in current code, in `writer.find_max_part()`.
```python
def find_max_part(row_groups):
    """
    Find the highest integer matching "**part.*.parquet" in referenced paths.
    """
    paths = [c.file_path or "" for rg in row_groups for c in rg.columns]
    # TODO: the following line should not be necessary
    paths = [p.decode() if isinstance(p, bytes) else p for p in paths]
```
I confirm that after the change in logic, previously related failing test cases are now ok.

martindurant replied:
Ah, this might be an unfortunate change then. Decoding the bytes can take a non-trivial amount of time when there are many columns, and in the typical read-only case, it's unnecessary.

yohplala commented:
Hello @martindurant ,
Wish you a happy new year!!!
Martin, please, do you have new feedbacks? (I am not willing to bother here, I can wait, but I think, even at small steps, it is better to continue than stopping for too long)
Thanks for your support! Bests,

martindurant commented on fastparquet/api.py:
```diff
@@ -371,6 +376,15 @@ def remove_row_groups(self, rgs, write_fmd:bool = True,
+            else:
+                # If generator, switch to list as this list is re-used.
+                rgs = list(rgs)
+        if not rgs:
```
Probably do this before the type checks, so that we don't end up trying to work with something like `[( )]` here.

yohplala replied:
Martin, your 1st comment is on the outadated code:
_Does the first branch ever get called? A tuple also has __iter__, actually many things do. If what you really want to check for generator, why not do exactly that (inspect.isgenerator) ?_

Please, check you review the last version, which reads:
```python
        if not isinstance(rgs, list):
            if isinstance(rgs, ThriftObject):
                rgs = [rgs]
            else:
                # Use `list()` here, not `[]`, as the latter does not transform
                # generator into list but encapsulates generator in a list.
                rgs = list(rgs)
```
(I confirm it is correctly loaded in github, this copy/paste is directly from github, in tab `files` of this repo)

This should answer your comment, the 'troublesome' case if when provided `rgs`is a simple `ThriftObect`. The other cases are correctly handled by `list()`, but not the case when `rgs` is a `ThriftObject`.

martindurant replied:
OK, got it. If `dict` is also acceptable for the condition (because a `ThriftObject` is just a wrapper for `dict`), please also allow that.

yohplala replied:
> OK, got it. If `dict` is also acceptable for the condition (because a `ThriftObject` is just a wrapper for `dict`), please also allow that.

I am sorry I don't understand.
`rgs` can be:
 - a row group (in which case it is a `ThriftObject`)
 - or an iterable of row groups (generator, list, tuple...)

I cannot imagine a case when a user would provide a `dict`, except if fastparquet, elsewhere, can also accept indifferently `dict`  as `ThriftObject`. Is it the case? If yes, then yes, let's also correctly manage this type with `[]`and not `list()` (on a `dict`, `list()` would only keep the keys of the dict, not encapsulate the dict in a list indeed)

yohplala replied:
> As you have noticed, fmd.row_groups returns a list of ThriftObjects with row-group type, but fmd[3] returns a list of dicts. Both have the same information. The mapping between field names and integer index is in the specs constant in cencoding.pyx. I don't know if we ever expect the user to manipulate the raw dicts, but the code in fastparquet does, i.e., those portions where I commented in this PR to keep the index notation rather than attribute access, because it is (much) faster.

Ok, clear, so I have added `idct` type checking as well.

martindurant commented:
The following comment arrived in my inbox apparently without a link to this PR, so pasting here.

>> OK, got it. If dict is also acceptable for the condition (because a ThriftObject is just a wrapper for dict), please also allow that.

> I am sorry I don't understand.
rgs can be:

> a row group (in which case it is a ThriftObject)
or an iterable of row groups (generator, list, tuple...)
I cannot imagine a case when a user would provide a dict, except if fastparquet, elsewhere, can also accept indifferently dict as ThriftObject. Is it the case? If yes, then yes, let's also correctly manage this type with []and not list() (on a dict, list() would only keep the keys of the dict, not encapsulate the dict in a list indeed)

As you have noticed, fmd.row_groups returns a list of ThriftObjects with row-group type, but fmd[3] returns a list of dicts. Both have the same information. The mapping between field names and integer index is in the `specs` constant in cencoding.pyx. I don't know if we ever expect the user to manipulate the raw dicts, but the code in fastparquet does, i.e., those portions where I commented in this PR to keep the index notation rather than attribute access, because it is (much) faster.

yohplala commented:
@martindurant 
I checked the dask failing test casess.
they are all related to a DeprecationWarning.
```bash
E       DeprecationWarning: the `interpolation=` argument to percentile was renamed to `method=`, which has additional options.
E       Users of the modes 'nearest', 'lower', 'higher', or 'midpoint' are encouraged to review the method they. (Deprecated NumPy 1.22)
```
This seems to be unrelated to this commit.

martindurant commented:
https://github.com/dask/dask/pull/8525

martindurant commented:
Am I right in thinking that the only thing left here is the questions of decoding all of the file paths on open, where this is costly in time and not needed for most users?

yohplala commented:
> Am I right in thinking that the only thing left here is the questions of decoding all of the file paths on open, where this is costly in time and not needed for most users?

@martindurant , I reverted this as well, this is the last message I wrote (decoding all of the file path occured in `api._parse_header`):

> Ok, sorry, I understand. I reverted the change, both in util.metadata_from_many() and in api._parse_header.

I took into account all your feedbacks.
For me 'all' is solved, waiting for new feedbacks ;)

yohplala commented:
Hi @martindurant ,
As indicated in previous message, all your feedbacks have been taken into account so far.
Please, if you have other ones to share, I can see to address them.
Thans! Bests,

yohplala commented:
Thanks @martindurant !
